File: chatkit-ui/src/main/java/com/netease/yunxin/kit/chatkit/ui/view/ScaleImageView.java
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

package com.netease.yunxin.kit.chatkit.ui.view;

import android.content.Context;
import android.graphics.Matrix;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import androidx.appcompat.widget.AppCompatImageView;

/** 自定义ImageView，支持缩放和拖动 */
public class ScaleImageView extends AppCompatImageView {

  private Matrix matrix = new Matrix();
  private ScaleGestureDetector scaleDetector;
  private float lastX, lastY;
  private boolean allPointersUp;
  // 支持缩放倍数
  private float scaleSize = 1f;

  public ScaleImageView(Context context) {
    this(context, null);
  }

  public ScaleImageView(Context context, AttributeSet attrs) {
    this(context, attrs, 0);
  }

  public ScaleImageView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context);
  }

  private void init(Context context) {
    scaleDetector = new ScaleGestureDetector(context, new ScaleListener());
    setScaleType(ScaleType.MATRIX);
  }

  public void centerImage() {
    postDelayed(
        () -> {
          reloadCenterImage();
          invalidate(); // 触发重绘
        },
        500);
  }

  private void reloadCenterImage() {
    Drawable drawable = getDrawable();
    if (drawable == null) return;

    // 获取图片和容器尺寸
    int imgWidth = drawable.getIntrinsicWidth();
    int imgHeight = drawable.getIntrinsicHeight();
    int viewWidth = getWidth();
    int viewHeight = getHeight();

    // 计算平移量
    float dx = (viewWidth - imgWidth) / 2f;
    float dy = (viewHeight - imgHeight) / 2f;

    matrix.postTranslate(dx, dy); // 应用初始平移
  }

  @Override
  public boolean onTouchEvent(MotionEvent event) {
    scaleDetector.onTouchEvent(event);
    final int action = event.getActionMasked();
    switch (action) {
      case MotionEvent.ACTION_DOWN:
        lastX = event.getX();
        lastY = event.getY();
        allPointersUp = false;
        break;
      case MotionEvent.ACTION_POINTER_DOWN:
        allPointersUp = true;
        break;
      case MotionEvent.ACTION_MOVE:
        if (!allPointersUp) {
          float dx = event.getX() - lastX;
          float dy = event.getY() - lastY;
          matrix.postTranslate(dx, dy);
          setImageMatrix(matrix);
          lastX = event.getX();
          lastY = event.getY();
        }
        break;
    }
    return true;
  }

  private class ScaleListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {
    @Override
    public boolean onScale(ScaleGestureDetector detector) {
      float scaleFactorChange = detector.getScaleFactor();
      // 获取ImageView的中心点作为缩放中心点
      float px = getWidth() / scaleSize;
      float py = getHeight() / scaleSize;
      // 在ImageView的中心点进行缩放
      matrix.postScale(scaleFactorChange, scaleFactorChange, px, py);
      // 应用新的矩阵变换
      setImageMatrix(matrix);
      return true;
    }

    @Override
    public void onScaleEnd(ScaleGestureDetector detector) {
      super.onScaleEnd(detector);
    }
  }
}

File: common-ui/src/main/java/com/netease/yunxin/kit/common/ui/widgets/scaleimage/ScaleImageView.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.yunxin.kit.common.ui.widgets.scaleimage

import android.content.Context
import android.graphics.Matrix
import android.graphics.RectF
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.AttributeSet
import android.view.GestureDetector
import androidx.appcompat.widget.AppCompatImageView

class ScaleImageView@JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int = 0
) : AppCompatImageView(context, attrs, defStyleAttr) {

    private lateinit var attache: ScaleAttache
    private var pendingScaleType: ScaleType? = null
    init {
        if (attrs != null) {
            attache = ScaleAttache(this)
            // We always pose as a Matrix scale type, though we can change to another scale type
            // via the attacher
            super.setScaleType(ScaleType.MATRIX)
            // apply the previously applied scale type
            if (pendingScaleType != null) {
                setScaleType(pendingScaleType!!)
                pendingScaleType = null
            }
        }
    }

    /**
     * Get the current [PhotoViewAttacher] for this view. Be wary of holding on to references
     * to this attacher, as it has a reference to this view, which, if a reference is held in the
     * wrong place, can cause memory leaks.
     *
     * @return the attacher.
     */
    fun getAttache(): ScaleAttache {
        return attache
    }

    override fun getScaleType(): ScaleType {
        return attache.getScaleType()
    }

    override fun getImageMatrix(): Matrix {
        return attache.getImageMatrix()
    }

    override fun setOnLongClickListener(l: OnLongClickListener?) {
        attache.setOnLongClickListener(l)
    }

    override fun setOnClickListener(l: OnClickListener?) {
        attache.setOnClickListener(l)
    }

    override fun setScaleType(scaleType: ScaleType) {
        if (attache == null) {
            pendingScaleType = scaleType
        } else {
            attache.setScaleType(scaleType)
        }
    }

    override fun setImageDrawable(drawable: Drawable?) {
        super.setImageDrawable(drawable)
        // setImageBitmap calls through to this method
        if (attache != null) {
            attache.update()
        }
    }

    override fun setImageResource(resId: Int) {
        super.setImageResource(resId)
        if (attache != null) {
            attache.update()
        }
    }

    override fun setImageURI(uri: Uri?) {
        super.setImageURI(uri)
        if (attache != null) {
            attache.update()
        }
    }

    override fun setFrame(l: Int, t: Int, r: Int, b: Int): Boolean {
        val changed = super.setFrame(l, t, r, b)
        if (changed) {
            attache.update()
        }
        return changed
    }

    fun setRotationTo(rotationDegree: Float) {
        attache.setRotationTo(rotationDegree)
    }

    fun setRotationBy(rotationDegree: Float) {
        attache.setRotationBy(rotationDegree)
    }

    fun isZoomable(): Boolean {
        return attache.isZoomable()
    }

    fun setZoomable(zoomable: Boolean) {
        attache.setZoomable(zoomable)
    }

    fun getDisplayRect(): RectF? {
        return attache.getDisplayRect()
    }

    fun getDisplayMatrix(matrix: Matrix) {
        attache.getDisplayMatrix(matrix)
    }

    fun setDisplayMatrix(finalRectangle: Matrix): Boolean {
        return attache.setDisplayMatrix(finalRectangle)
    }

    fun getSuppMatrix(matrix: Matrix) {
        attache.getSuppMatrix(matrix)
    }

    fun setSuppMatrix(matrix: Matrix?): Boolean {
        return attache.setDisplayMatrix(matrix)
    }

    fun getMinimumScale(): Float {
        return attache.getMinimumScale()
    }

    fun getMediumScale(): Float {
        return attache.getMediumScale()
    }

    fun getMaximumScale(): Float {
        return attache.getMaximumScale()
    }

    fun getScale(): Float {
        return attache.getScale()
    }

    fun setAllowParentInterceptOnEdge(allow: Boolean) {
        attache.setAllowParentInterceptOnEdge(allow)
    }

    fun setMinimumScale(minimumScale: Float) {
        attache.setMinimumScale(minimumScale)
    }

    fun setMediumScale(mediumScale: Float) {
        attache.setMediumScale(mediumScale)
    }

    fun setMaximumScale(maximumScale: Float) {
        attache.setMaximumScale(maximumScale)
    }

    fun setScaleLevels(minimumScale: Float, mediumScale: Float, maximumScale: Float) {
        attache.setScaleLevels(minimumScale, mediumScale, maximumScale)
    }

    fun setOnMatrixChangeListener(listener: OnMatrixChangedListener?) {
        attache.setOnMatrixChangeListener(listener)
    }

    fun setOnPhotoTapListener(listener: OnPhotoTapListener?) {
        attache.setOnPhotoTapListener(listener)
    }

    fun setOnOutsidePhotoTapListener(listener: OnOutsidePhotoTapListener?) {
        attache.setOnOutsidePhotoTapListener(listener)
    }

    fun setOnViewTapListener(listener: OnViewTapListener?) {
        attache.setOnViewTapListener(listener)
    }

    fun setOnViewDragListener(listener: OnViewDragListener?) {
        attache.setOnViewDragListener(listener)
    }

    fun setScale(scale: Float) {
        attache.setScale(scale)
    }

    fun setScale(scale: Float, animate: Boolean) {
        attache.setScale(scale, animate)
    }

    fun setScale(scale: Float, focalX: Float, focalY: Float, animate: Boolean) {
        attache.setScale(scale, focalX, focalY, animate)
    }

    fun setZoomTransitionDuration(milliseconds: Int) {
        attache.setZoomTransitionDuration(milliseconds)
    }

    fun setOnDoubleTapListener(onDoubleTapListener: GestureDetector.OnDoubleTapListener?) {
        attache.setOnDoubleTapListener(onDoubleTapListener)
    }

    fun setOnScaleChangeListener(onScaleChangedListener: OnScaleChangedListener?) {
        attache.setOnScaleChangeListener(onScaleChangedListener)
    }

    fun setOnSingleFlingListener(onSingleFlingListener: OnSingleFlingListener?) {
        attache.setOnSingleFlingListener(onSingleFlingListener)
    }
}
